### ISSUE TITLE
[DISCUSS] Structural changes to ResponseParser

### ISSUE THREAD

**Issue by nevans** [ruby/net-imap#11]
I would like to split `Net::IMAP::ResponseParser` into three parts: the parser, the lexer, and a "builder" (so the data structures built by the parser can be more easily customized). **Would a patch that does just this (and nothing else) be accepted?**

The parser seems to have been based on an obsolete IMAP version. E.g. some of the comments and code around astrings (see fe77596) match the older [RFC2060]. Although I can continue creating smaller patches to update smaller pieces, **would a much larger cleanup project to more closely match RFC3501 be accepted?**

And, **what do you think about rewriting the parser with racc?** The lexer could be converted to `StringScanner` or to `ragel` or it could keep the same `#next_token` it has now. I have some improvements on diagnostics for parse errors that I'd like to add as well.

**Would you accept big patches to build out a _much_ larger `ResponseParser` test suite?** This should include benchmark scripts to protect against regressions. This should be a prerequisite for the bigger updates, like `racc` or `StringScanner` (or `ragel`).

My company processes GBs of IMAP every day, and speeding up `ResponseParser` would be a godsend. CPU is the limiting resource for my IMAP daemons, and I've wanted to rewrite or replace our parser for years now. For our newest project, my team has been using golang instead of ruby, even though the go IMAP library had fewer features and is more difficult to work with. That makes me so sad. :(

So in summary, are PRs welcome to:
* #10 Split `ResponseParser` into multiple components:
* Extract a lexer and "builder" from the existing ResponseParser class..
* A big update of the existing recursive descent parser to match [RFC3501] and [RFC4466] (and others).
* Greatly expand the ResponseParser test suite?.
* Convert the parser to `racc`?
* Convert the lexer to `StringScanner`?
  * _(maybe)_ Convert the lexer to ragel-generated `C` code?

Thanks!

[RFC2060]: https://tools.ietf.org/html/rfc2060
[RFC3501]: https://tools.ietf.org/html/rfc3501
[RFC4466]: https://tools.ietf.org/html/rfc4466

**Comment by shugo** [comment]
@nevans Your proposal sounds fine.

I would like you to have write access to the repository.
What do you think?

**Comment by nevans** [comment]
@shugo Sorry, I somehow missed your comment on this.

I would be honored to have write access! I'll still want a second pair of eyes on my work before I merge them into master, of course. 😄  Thanks!

**Comment by shugo** [comment]
@nevans Thank you!  We've invited you to the repository.

**Comment by nevans** [comment]
I'm closing this ticket, as it was just meant to get your feedback on my proposal.  I'll open a new tickets for the various tasks when I start on them.  :)
